From fb1964f6c90940ad1808010cf073b53c49a76205 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:40:35 -0400
Subject: [alsa] hda: Fix probing and stuttering on CMI8888 HD-audio controller

Message-id: <140972823573-413-patchset-jkysela@redhat.com>
Patchwork-id: 91109
O-Subject: [RHEL 7.1 PATCH 412/442] ALSA: hda - Fix probing and stuttering on CMI8888 HD-audio controller
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

ASUS Phoebus with CMI8888 HD-audio chip (PCI id 13f6:5011) doesn't
work with HD-audio driver as is because of some weird nature.  For
making DMA properly working, we need to disable MSI.  The position
report buffer doesn't work, thus we need to force reading LPIB
instead.  And yet, the codec CORB/RIRB communication gives errors
unless we disable the snooping (caching).

In this patch, all these workarounds are added as a quirk for the
device.  The HD-audio *codec* chip needs yet another workaround, but
it'll be provided in the succeeding patch.

Reported-and-tested-by: Vincent Lejeune <vljn@ovi.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: c563f473ac31c182d93f4c6ec8eb0ce199439978)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/hda_intel.c b/sound/pci/hda/hda_intel.c
index 390ef33..2103316 100644
--- a/sound/pci/hda/hda_intel.c
+++ b/sound/pci/hda/hda_intel.c
@@ -265,6 +265,7 @@ enum {
  AZX_DRIVER_TERA,
  AZX_DRIVER_CTX,
  AZX_DRIVER_CTHDA,
+ AZX_DRIVER_CMEDIA,
  AZX_DRIVER_GENERIC,
  AZX_NUM_DRIVERS, /* keep this as last entry */
 };
@@ -330,6 +331,7 @@ static char *driver_short_names[] = {
  [AZX_DRIVER_TERA] = "HDA Teradici", 
  [AZX_DRIVER_CTX] = "HDA Creative", 
  [AZX_DRIVER_CTHDA] = "HDA Creative",
+ [AZX_DRIVER_CMEDIA] = "HDA C-Media",
  [AZX_DRIVER_GENERIC] = "HD-Audio Generic",
 };
 
@@ -1373,6 +1375,7 @@ static void azx_check_snoop_available(struct azx *chip)
   snoop = false;
   break;
  case AZX_DRIVER_CTHDA:
+ case AZX_DRIVER_CMEDIA:
   snoop = false;
   break;
  }
@@ -2154,6 +2157,10 @@ static const struct pci_device_id azx_ids[] = {
    .driver_data = AZX_DRIVER_CTX | AZX_DCAPS_CTX_WORKAROUND |
    AZX_DCAPS_RIRB_PRE_DELAY | AZX_DCAPS_POSFIX_LPIB },
 #endif
+ /* CM8888 */
+ { PCI_DEVICE(0x13f6, 0x5011),
+   .driver_data = AZX_DRIVER_CMEDIA |
+   AZX_DCAPS_NO_MSI | AZX_DCAPS_POSFIX_LPIB },
  /* Vortex86MX */
  { PCI_DEVICE(0x17f3, 0x3010), .driver_data = AZX_DRIVER_GENERIC },
  /* VMware HDAudio */
-- 
1.7.1